syntax = "proto3";

package endpoints.marketCentralPlatform;

option java_package = "com.google.endpoints.marketplace";
option java_outer_classname = "MarketCentralPlatformProto";
option java_multiple_files = true;

// Market Central Platform service
service MarketCentralPlatform {
    // Seller
    rpc sellerRegisteration(sellerRegisterationRqst) returns (sellerRegisterationResp);
    rpc itemSold(itemSoldRqst) returns (itemSoldResp);
    rpc itemUpdated(itemUpdatedRqst) returns (itemUpdatedResp);
    rpc itemRemoved(itemRemovedRqst) returns (itemRemovedResp);
    rpc itemsDisplayed(itemsDisplayedRqst) returns (itemsDisplayedResp);

    // Buyer
    rpc buyerRegisteration(buyerRegisterationRqst) returns (buyerRegisterationResp);
    rpc itemSearch(itemSearchRqst) returns (itemSearchResp);
    rpc itemBought(itemBoughtRqst) returns (itemBoughtResp);
    rpc itemWishlisted(itemWishlistedRqst) returns (itemWishlistedResp);
    rpc itemRated(itemRatedRqst) returns (itemRatedResp);
    
    // RPC Notification 
    rpc clientNotified(clientNotifiedRqst) returns (clientNotifiedResp);
}

// Resources & RPC Requests and Responses

message item {
    string id = 1;
    string name = 2;
    float price = 3;
    int32 quantity = 4;
    enum categories {
        ELECTRONICS = 0;
        FASHION = 1;
        OTHERS = 2;
    }
    categories category = 5;
    string description = 6;
    float rating = 7;
    string seller_id = 8;
    string seller_address = 9;
    string verbose = 10;
}

message sellerClient {
    string id = 1;
    string address = 2;
    repeated item items = 3;
    string verbose = 4;
}

message buyerClient {
    string id = 1;
    string address = 2;
    repeated item wishlist = 3;
    string verbose = 4;
}

message sellerRegisterationRqst {
    string id = 1;
    string address = 2;
    string verbose = 3;
}

message sellerRegisterationResp {
    string message = 1;
    string verbose = 2;
}

message itemSoldRqst {
    string id = 1;
    string name = 2;
    float price = 3;
    int32 quantity = 4;
    enum categories {
        ELECTRONICS = 0;
        FASHION = 1;
        OTHERS = 2;
    }
    categories category = 5;
    string description = 6;
    string seller_id = 7;
    string seller_address = 8;
    string verbose = 9;
}

message itemSoldResp {
    string message = 1;
    string verbose = 2;
}

message itemUpdatedRqst {
    string id = 1;
    float price = 2;
    int32 quantity = 3;
    string seller_id = 4;
    string seller_address = 5;
    string verbose = 6;
}

message itemUpdatedResp {
    string message = 1;
    string verbose = 2;
}

message itemRemovedRqst {
    string id = 1;
    string seller_id = 2;
    string seller_address = 3;
    string verbose = 4;
}

message itemRemovedResp {
    string message = 1;
    string verbose = 2;
}

message itemsDisplayedRqst {
    string seller_id = 1;
    string seller_address = 2;
    string verbose = 3;
}

message itemsDisplayedResp {
    string message = 1;
    string verbose = 2;
}

message buyerRegisterationRqst {
    string id = 1;
    string address = 2;
    string verbose = 3;
}

message buyerRegisterationResp {
    string message = 1;
    string verbose = 2;
}

message itemSearchRqst {
    string buyer_id = 1;
    string buyer_address = 2;
    string name = 3;
    enum categories {
        ELECTRONICS = 0;
        FASHION = 1;
        OTHERS = 2;
        ANY = 3;
    }
    categories category = 4;
    string verbose = 5;
}

message itemSearchResp {
    string message = 1;
    string verbose = 2;
}

message itemBoughtRqst {
    string item_id = 1;
    int32 item_quantity = 2;
    string buyer_id = 3;
    string buyer_address = 4;
    string verbose = 5;
}

message itemBoughtResp {
    string message = 1;
    string verbose = 2;
}

message itemWishlistedRqst {
    string item_id = 1;
    string buyer_id = 2;
    string buyer_address = 3;
    string verbose = 4;
}

message itemWishlistedResp {
    string message = 1;
    string verbose = 2;
}

message itemRatedRqst {
    string item_id = 1;
    float rating = 2;
    string buyer_id = 3;
    string buyer_address = 4;
    string verbose = 5;
}

message itemRatedResp {
    string message = 1;
    string verbose = 2;
}

message clientNotifiedRqst {
    string item_id = 1;
    enum c_type{
        SELLER = 0;
        BUYER = 1;
    }
    c_type client_type = 2;
    string verbose = 3;
}

message clientNotifiedResp {
    string message = 1;
    string verbose = 2;
}